src/log_rate.c

```diff
@@ -0,0 +1,110 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "utils.h"
+#include "log_rate.h"
+#include "ctr_logging.h"
+
+#define IO_BUF_SIZE      65536
+#define BYTES_PER_PERIOD 1024
+#define SECS_PER_PERIOD  1
+#define BILLION          1000000000
+
+static const struct timespec secs_per_period = {SECS_PER_PERIOD, 0};
+static size_t bytes_written_this_period = 0;
+static struct timespec start_of_this_period;
+
+static struct timespec subtract_timespecs(const struct timespec* first, const struct timespec* second);
+static int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second);
+static void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count);
+static void sleep_for_the_rest_of_this_period();
+static void start_new_period();
+
+void log_rate_init() {
+	start_new_period();
+}
+
+bool log_rate_write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read) {
+	char* buf_start = buf;
+	ssize_t bytes_remaining = num_read;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	int64_t diff_nano = subtract_timespecs_nano(&now, &start_of_this_period);
+	if (diff_nano < SECS_PER_PERIOD * BILLION) {
+		ssize_t bytes_we_can_write = BYTES_PER_PERIOD - bytes_written_this_period;
+		if (num_read <= bytes_we_can_write) {
+			write_io_bufs(pipe, buf_start, num_read);
+			bytes_written_this_period += num_read;
+			return true;
+		} else {
+			write_io_bufs(pipe, buf_start, bytes_we_can_write);
+			bytes_written_this_period += bytes_we_can_write;
+			buf_start += bytes_we_can_write;
+			bytes_remaining = num_read - bytes_we_can_write;
+			sleep_for_the_rest_of_this_period();
```

Comment:
I think we shouldn't block the thread execution here.

We should either propagate the condition back (with the remaining time) and set a glibc timer, or we need to use a much shorter period so that it is acceptable, in the worst case, to hang for that duration

Reply:
This code implements the backpressure policy, I'm still adding the others. 
I'm not sure I follow your concern or how backpressure is going to be maintained if we return from here without blocking. If we do, aren't we going to quickly enough be reading from the container's pipe again?

Reply:
My first reaction to the sleep was also "yeuch" but given that conmon is written as a sync read()/write() loop there's no other way to idle but to sleep.

If conmon was written as a poll/epoll loop then you could set a timer and flip the read fd out of the loop till it fired. That would be neater and more efficient, but a bigger overhaul of conmon.